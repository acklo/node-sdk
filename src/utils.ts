import fs from "fs";
import path from "path";

export async function delay(ms: number): Promise<void> {
  return new Promise<void>((resolve) => {
    setTimeout(() => resolve(), ms);
  });
}

function isFile(path: string): boolean {
  return fs.existsSync(path) && !fs.lstatSync(path).isDirectory();
}

// Based on Jest's implementation for finding the jest.config file:
// https://github.com/facebook/jest/blob/master/packages/jest-config/src/resolveConfigPath.ts
export function findFile(
  filename: string | string[],
  pathToResolve: string
): string | false {
  const filenames = Array.isArray(filename) ? filename : [filename];

  const resolvedPath = filenames
    .map((p) => path.resolve(pathToResolve, p))
    .find(isFile);

  if (resolvedPath) {
    return resolvedPath;
  }

  if (pathToResolve === path.dirname(pathToResolve)) {
    return false;
  }

  return findFile(filenames, path.dirname(pathToResolve));
}
