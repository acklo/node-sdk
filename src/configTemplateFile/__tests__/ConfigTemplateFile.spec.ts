import fs from "fs";
import {
  ConfigTemplateFile,
  ConfigTemplateEventKind,
  ConfigTemplateValueUpdates,
} from "../ConfigTemplateFile";
import { delay } from "../../utils";

describe("ConfigTemplateFile", () => {
  let configTemplate: ConfigTemplateFile;

  beforeAll(() => {
    const yaml = fs
      .readFileSync(
        "./src/configTemplateFile/__tests__/configTemplateExample.yml"
      )
      .toString("utf-8");
    configTemplate = new ConfigTemplateFile(yaml);
  });

  it("converts json string to a ConfigTemplate", () => {
    expect(configTemplate.content).toMatchSnapshot();
  });

  it("returns the defaults from the json config when no other values are available", () => {
    expect(configTemplate.values()).toMatchInlineSnapshot(`
      Object {
        "app_config.log_level": "trace",
        "app_config.port": 3000,
        "feature_switches.motd": "Be good to each other.",
        "feature_switches.new_header": false,
      }
    `);
  });

  it("can update the config template's values", () => {
    expect(configTemplate.updateValues({})).toEqual({});

    configTemplate.updateValues({
      "app_config.log_level": "info",
      "feature_switches.motd": "Happy Monday!",
      "feature_switches.new_header": "true",
      "app_config.port": "3000",
    });

    expect(configTemplate.values()).toMatchInlineSnapshot(`
      Object {
        "app_config.log_level": "info",
        "app_config.port": 3000,
        "feature_switches.motd": "Happy Monday!",
        "feature_switches.new_header": true,
      }
    `);

    configTemplate.updateValues({
      "feature_switches.new_header": "false",
    });

    expect(configTemplate.values()).toMatchInlineSnapshot(`
      Object {
        "app_config.log_level": "info",
        "app_config.port": 3000,
        "feature_switches.motd": "Happy Monday!",
        "feature_switches.new_header": false,
      }
    `);
  });

  it("emits an event when updates come through", async () => {
    const updates: ConfigTemplateValueUpdates[] = [];
    configTemplate.events.on(ConfigTemplateEventKind.ConfigUpdate, (update) => {
      updates.push(update);
    });

    configTemplate.updateValues({
      "app_config.log_level": "debug",
      "app_config.port": "3001",
    });

    configTemplate.updateValues({
      "app_config.log_level": "info",
    });

    await delay(10);

    expect(updates).toMatchInlineSnapshot(`
      Array [
        Object {
          "app_config.log_level": Object {
            "newValue": "debug",
            "oldValue": "info",
          },
          "app_config.port": Object {
            "newValue": 3001,
            "oldValue": 3000,
          },
        },
        Object {
          "app_config.log_level": Object {
            "newValue": "info",
            "oldValue": "debug",
          },
        },
      ]
    `);
  });

  describe("when receiving an update for a property that does not exist in the config template", () => {
    const updates: ConfigTemplateValueUpdates[] = [];

    beforeAll(async () => {
      configTemplate.events.on(
        ConfigTemplateEventKind.ConfigUpdate,
        (update) => {
          updates.push(update);
        }
      );

      configTemplate.updateValues({
        "app_config.yolo": "true",
      });

      await delay(10);
    });

    it("does not emit any updates", () => {
      expect(updates).toHaveLength(0);
    });

    it("does not include the property when retrieving values", () => {
      expect(configTemplate.values()["app_config.yolo"]).toBeUndefined();
    });
  });

  it("returns the config as JSON", () => {
    expect(configTemplate.toJSON()).toMatchSnapshot();
  });
});
