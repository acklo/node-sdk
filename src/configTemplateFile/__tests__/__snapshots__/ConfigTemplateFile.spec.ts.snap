// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConfigTemplateFile converts json string to a ConfigTemplate 1`] = `
Object {
  "configuration": Array [
    Object {
      "description": "Key configuration for the app.",
      "id": "app_config",
      "name": "App config",
      "properties": Array [
        Object {
          "default": "trace",
          "id": "log_level",
          "name": "Log level",
          "type": "string",
        },
        Object {
          "default": 3000,
          "id": "port",
          "name": "Server port",
          "type": "number",
        },
      ],
    },
    Object {
      "description": "Feature switches for the app.",
      "id": "feature_switches",
      "name": "Feature switches",
      "properties": Array [
        Object {
          "default": "Be good to each other.",
          "id": "motd",
          "name": "Message of the day",
          "type": "string",
        },
        Object {
          "default": false,
          "id": "new_header",
          "name": "Show new header",
          "type": "boolean",
        },
      ],
    },
  ],
  "version": "v1",
}
`;

exports[`ConfigTemplateFile returns the config as JSON 1`] = `
"{
  \\"configuration\\": [
    {
      \\"description\\": \\"Key configuration for the app.\\",
      \\"id\\": \\"app_config\\",
      \\"name\\": \\"App config\\",
      \\"properties\\": [
        {
          \\"default\\": \\"trace\\",
          \\"id\\": \\"log_level\\",
          \\"name\\": \\"Log level\\",
          \\"type\\": \\"string\\"
        },
        {
          \\"default\\": 3000,
          \\"id\\": \\"port\\",
          \\"name\\": \\"Server port\\",
          \\"type\\": \\"number\\"
        }
      ]
    },
    {
      \\"description\\": \\"Feature switches for the app.\\",
      \\"id\\": \\"feature_switches\\",
      \\"name\\": \\"Feature switches\\",
      \\"properties\\": [
        {
          \\"default\\": \\"Be good to each other.\\",
          \\"id\\": \\"motd\\",
          \\"name\\": \\"Message of the day\\",
          \\"type\\": \\"string\\"
        },
        {
          \\"default\\": false,
          \\"id\\": \\"new_header\\",
          \\"name\\": \\"Show new header\\",
          \\"type\\": \\"boolean\\"
        }
      ]
    }
  ],
  \\"version\\": \\"v1\\"
}"
`;
